doc/sonic-application-extention/sonic-application-extention-hld.md

```diff
@@ -0,0 +1,1441 @@
+| /service/dependent-of        | lits of strnigs | no        | List of SONiC services this application is dependent of.<p>Specifying in this option a service X, will regenerate the /usr/local/bin/X.sh script and upgrade the "DEPENDENT" list with this package service.<p>This option is warm-restart related, a warm-restart of service X will not trigger this package service restart.<p>On the other hand, this service package will be started, stopped, restarted togather with service X.<p>Example:<p>For "dhcp-relay", "radv", "teamd" this field will have "swss" service in the list. |
+| /service/post-start-action   | string          | no        | Path to an executable inside Docker image filesystem to be executed after container start.<p>A package may use this field in case a systemd service should not reach started state before some condition. E.g.: A database service should not reach started state before redis process is not ready. Since, there is no control, when the redis process will start a "post-start-action" script may execute "redis-cli ping" till the ping is succeessful.                                                                            |
+| /service/pre-shutdown-action | string          | no        | Path to an executable inside Docker image filesystem to be executed before container stops.<p>A uses case is to execute a warm-shutdown preparation script.<p>A script that sends SIGUSR1 to teamd to initiate warm shutdown is one of such examples.                                                                                                                                                                                                                                                                                 |
+
+<!-- omit in toc -->
+##### /usr/bin/*feature*.sh
+
+The script under /usr/bin/ starts, stops or waits for container exit. This script is auto-generated during build time from
+[docker_image_ctl.j2](https://github.com/Azure/sonic-buildimage/blob/4006ce711fa6545b0870186ffa05d4df24edb8b7/files/build_templates/docker_image_ctl.j2).
+To allow a runtime package installation, it is required to have this file as part of SONiC image and put it in
+*/usr/share/sonic/templates/docker_image_ctl.j2*. The Jinja2 template will accept three arguments, *docker_container_name*,
+*docker_image_name* and *docker_run_options*, which derive from the */container/* node from manifest. Besides of options
+defined in the manifest, the following default are used to start container to allow container to access base SONiC resources,
+like database and syslog:
+
+```
+docker create {{ docker_run_options }}
+  --net=$NET \
+  --uts=host \
+  --log-opt max-size=2M --log-opt max-file=5 \
+  -v /var/run/redis$DEV:/var/run/redis:rw \
+  $REDIS_MNT \
+  -v /usr/share/sonic/device/$PLATFORM:/usr/share/sonic/platform:ro \
+  -v /usr/share/sonic/device/$PLATFORM/$HWSKU/$DEV:/usr/share/sonic/hwsku:ro \
```

Comment:
We may not need these mount volumes for 3rd party dockers as they may not depend on SONiC resources? can we make it an optional field (e.g SONiC dependent docker vs non-dependent docker) in the manifest file?

Reply:
This is mentioned in https://github.com/Azure/SONiC/pull/682/files#diff-be6a02b89b88eb9dd23104151d3b06081dbc49c58c610f5f2aa5dfc5d1d4e13aR1298.

Normally, if you don't need these volume mounts you will not specify "depends": ["database"], or any other sonic container that has a dependency on database. In this case sonic-package-manager will know that these mounts are not needed.

Please, note, 3rd party dockers are not part of phase 1 implementation

Reply:
ok, thanks.